### PR TITLE
Do storeKey conversion for keys in GET/DELETE requests

### PR DESCRIPTION
**Pull request by zzmao**
@dharju Looks like we don't need a mock `StoreKeyConverterFactory` for test. `StoreKeyConverterImplNoOp` with existing tests can cover the new added method `getConvertedStoreKeys`.

### CODE REVIEW

**Comment by vgkholla** [comment]
@zzmao What tests would fail if i removed the `getConvertedKeys()` call?

**Comment by zzmao** [comment]
Nothing fail without `getConvertedStoreKeys()`. 

`getConvertedStoreKeys` 's convertion doesn't change any key because tests use `StoreKeyConverterImplNoOp `

**Comment by vgkholla** [comment]
@zzmao Yeah - so if the code is "incorrect" without the call (which it will be), then there should be a test that fails if that call is removed

**Comment by dharju** [comment]
Yeah, like @vgkholla said we should have a test that fails without the storekeyconverter change and succeeds when it is made.  That will likely need a MockStoreKeyConverter to do.

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 1]
```diff
@@ -393,16 +395,17 @@ public void handleDeleteRequest(Request request) throws IOException, Interrupted
...
       if (error != ServerErrorCode.No_Error) {
         logger.error("Validating delete request failed with error {} for request {}", error, deleteRequest);
         response = new DeleteResponse(deleteRequest.getCorrelationId(), deleteRequest.getClientId(), error);
       } else {
         MessageFormatInputStream stream =
-            new DeleteMessageFormatInputStream(deleteRequest.getBlobId(), deleteRequest.getAccountId(),
+            new DeleteMessageFormatInputStream(convertedStoreKey, deleteRequest.getAccountId(),
                 deleteRequest.getContainerId(), deleteRequest.getDeletionTimeInMs());
-        MessageInfo info = new MessageInfo(deleteRequest.getBlobId(), stream.getSize(), deleteRequest.getAccountId(),
+        MessageInfo info = new MessageInfo(convertedStoreKey, stream.getSize(), deleteRequest.getAccountId(),
```
Comment (vgkholla) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: i think the notification system below also needs to use the converted ID

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java [hunk 2]
```diff
@@ -1061,4 +1064,16 @@ private boolean isRemoteLagLesserOrEqual(Collection<PartitionId> partitionIds, l
...
     return isAcceptable;
   }
+
+  /**
+   * Convert StoreKeys based on {@link StoreKeyConverterFactory}
+   * @param storeKeys A list of original storeKeys.
+   * @return A list of converted storeKeys.
+   */
+  private List<StoreKey> getConvertedStoreKeys(List<? extends StoreKey> storeKeys) throws Exception {
+    Map<StoreKey, StoreKey> convertionMap = storeKeyConverterFactory.getStoreKeyConverter().convert(storeKeys);
```
Comment (vgkholla) [inline ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: typo: conversion

File: ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java [hunk 3]
```diff
@@ -127,8 +129,11 @@ public AmbryRequestsTest() throws IOException, ReplicationException, StoreExcept
     dataNodeId = clusterMap.getDataNodeIds().get(0);
     replicationManager =
         MockReplicationManager.getReplicationManager(verifiableProperties, storageManager, clusterMap, dataNodeId);
+    StoreKeyConverterFactory storeKeyConverterFactory =
```
Comment (vgkholla) [inline ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: will need a mock to check functionality. I think you can write a mock in in ambry-api package (in the tests module). Something that will let you add and remove mappings from the converter either as individual mappings or an input hashmap.

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java (hunk 2 above)
Comment (zzmao) [reply ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: fixed.

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java (hunk 1 above)
Comment (zzmao) [reply ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: Or should we note down both? 
I though we/customer may be interested in the original key's event.

File: ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java (hunk 3 above)
Comment (zzmao) [reply ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: Implements `StoreKeyConverterFactory/StoreKeyConverter` in ambry-api  is a little difficulty. They need the context in the test. 

I added them in `AmbryRequestsTest.java`, looks OK.

**Comment by zzmao** [comment]
Tests added. Now test failed without the conversion in prod code.

**Comment by zzmao** [comment]
Addressed @vgkholla comment.

File: ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java [hunk 4]
```diff
@@ -1053,9 +1156,12 @@ public void shutdown() throws StoreException {
      */
     PartitionId startedPartitionId = null;
 
-    MockStorageManager() throws StoreException {
+    private final Set<StoreKey> validKeysInStore;
```
Comment (vgkholla) [inline ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: minor: would it be easier/better to store a list of invalid keys?

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java (hunk 1 above)
Comment (vgkholla) [reply ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: Good point - I think it should be the converted key but let's chat.

File: ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java (hunk 4 above)
Comment (zzmao) [reply ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: The mock store impl needs to check if a key in `validKeysInStore` or not. Set is faster.

Comment (vgkholla) [reply ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: Ah set is fine - but I was wondering about what you want to track - valid or invalid keys?

Comment (vgkholla) [reply ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: Here the assumption is that a key is invalid if it is not explicitly present in `validKeysInStore` - the other approach is that it is invalid only if it is present in `invalidKeysInStore`

Comment (zzmao) [reply ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: I feel `validKeysInStore` is more natural. It's like keys are in real store. 
Plus, `invalidKeysInStore ` needs to add key everytime we want to check. Code maybe not friendly.

File: ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java [hunk 5]
```diff
@@ -528,6 +538,53 @@ public void stopBlobStoreFailureTest() throws InterruptedException, IOException
...
+
+  /**
+   * Tests blobIds can be converted as expected and works correctly with GetRequest.
+   * If all blobIds can be converted correctly, no error is expected.
+   * If any blobId can't be converted correctly, Blob_Not_Found is expected.
+   * @throws InterruptedException
+   * @throws IOException
+   */
+  @Test
+  public void listOfOriginalStoreKeysGetTest() throws Exception {
```
Comment (dharju) [inline ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: Is there an explicit deletion test?  I'm not sure if I'm seeing it

Comment (vgkholla) [reply ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: There isn't - only the integration tests test that unfortunately (this class didn't have any tests until a few months ago)

Comment (zzmao) [reply ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: It's covered by existing test `controlRequestSuccessTest`. Without the conversion, delete fails. 

`GetRequest` may request multiple keys, so `listOfOriginalStoreKeysGetTest ` was added for enhancement.

Comment (dharju) [reply ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: Ok cool, so all new code has coverage (when run through "Run * with coverage")?

Comment (zzmao) [reply ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: Yes. New changes covered.

File: ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java (hunk 1 above)
Comment (zzmao) [reply ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: @dharju What's your opinion about the key for notification?

Comment (dharju) [reply ambry-server/src/main/java/com.github.ambry.server/AmbryRequests.java]: Where does this notification go exactly?  Does it go to the client ultimately?

File: ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java [hunk 6]
```diff
@@ -617,6 +674,37 @@ private void doRequestControlRequestTest(RequestOrResponseType toControl, Partit
...
+        new GetRequest(correlationId, clientId, MessageFormatFlags.All, Collections.singletonList(pRequestInfo),
+            GetOption.Include_All);
+    storageManager.resetStore();
+    if (!expectedErrorCode.equals(ServerErrorCode.Unknown_Error)) {
+      // known error will be filled to each PartitionResponseInfo and set ServerErrorCode.No_Error in response.
+      Response response = sendRequestGetResponse(request, ServerErrorCode.No_Error);
+      assertEquals("Operation received at the store not as expected", RequestOrResponseType.GetRequest,
+          MockStorageManager.operationReceived);
+      for (PartitionResponseInfo info : ((GetResponse) response).getPartitionResponseInfoList()) {
+        assertEquals("Error code does not match expected", expectedErrorCode, info.getErrorCode());
```
Comment (vgkholla) [inline ambry-server/src/test/java/com.github.ambry.server/AmbryRequestsTest.java]: would it also be useful to check what keys you got in the store? How do you know you got the right converted key?

I'll merge this for now but let's make this change in another PR if required

**Pull request merged**